### DISCUSSION TITLE
Chapter 1 - Class in JS

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#4] [Chapter1]
# 3. Class in JS

JS has a class like other programing languages do. It's kind of synthetic sugar. It's contained from newer ECMA Script. Actually It's not necessarily understanding class in JS. So I will tell you key points only. 

Key points of class in JS
 - constructor only exist one.
 - constructor can't be assigned with arrow function.
 - ~~there is no property (field) declarations. you can assign it from inside of constructor.~~ *check out comment*
 - all properties and functions are public unless type prefix `#`.
 - `getter` and `setter` can be declared using `get` and `set`
 ```
ex)
get getSomething() {
.
.
.
  return property1;
}

set setSomething(a) {
.
.
.
  this.property1 = a;
}
```
 - `getter` and `setter` can't be called like a `function`.
```
ex)
const test = a.getSomething(); -> wrong way
const test = a.getSomething; -> correct way
a.setSomething(b); -> wrong way
a.setSomething = b; -> correct way
```
 - JS class can get static function too. add `static` before function name.
 
## ***That's all. It's not important in React I think. Just read it and go ahead.***

**Comment by honaseung** [comment]
Besides, one thing good to know.
`Object.getPrototypeOf(someObj);`
This can help you to look inside of class. You can see constructor and methods only.

**Comment by honaseung** [comment]
'There is no property (field) declarations. you can assign it from inside of constructor.' is wrong

**Reply by chaminju927** [reply]
I was curious about that as well ! please let me know the reason why that is wrong .

**Reply by honaseung** [reply]
You can declare or assign property inside of class scope level. You don't have to wait typing inside of constructor. But still you have to get through constructor if you want to save specific values as properties. 😂

**Comment by chaminju927** [comment]
"all properties and functions are public unless type prefix #. " -> this means that using prefix "#" makes the function private, right?

**Reply by honaseung** [reply]
Yeah, You are right. Let me show you an example.

```
a = 'test';
#b = 'test2';
```

Now a is a public property. And b is a private property. 😀
